python/ray/_raylet.pyx

```diff
@@ -2244,6 +2245,10 @@ cdef execute_task_with_cancellation_handler(
         # Check for cancellation.
         PyErr_CheckSignals()
 
+        # Register cleanup callback after actor is fully initialized (including __init__)
+        if <int>task_type == <int>TASK_TYPE_ACTOR_CREATION_TASK:
+            _register_actor_shutdown_callback()
```

Comment:
still not sure why we are lazily registering the callback. The C++ core worker logic knows when the actor has been initialized successfully and can call the callback only if that happens. There's no need to implicitly enforce that by registering the callback later.